packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -87,26 +394,58 @@ window.define = function(deps: string[], factory?: ResolveCallback) {
     if (defined === false) {
       pendingDefine = undefined;
       const url = baseUrl + '#' + topLevelScriptIdx++ as NormalizedUrl;
-      const mod = getModule(url);
+      // It's actually Initialized, but we're skipping over the Loading
+      // state, because this is a top level document and it's already loaded.
+      const mod = getModule(url) as Module<Loading>;
 
       // Top-level scripts are already loaded.
       mod.isTopLevel = true;
-      mod.needsLoad = false;
-
-      if (previousTopLevelUrl !== undefined) {
+      const predecessor = previousTopLevelUrl;
+      previousTopLevelUrl = url;
+      const waitingModule = beginWaitingOnEarlierScripts(mod, deps, factory);
+      const nextStep = () => {
+        beginWaitingOnDeps(waitingModule);
+        function failureShouldThrow(module: Module) {
```

Comment:
could this logic move into the fail function? there is already a `topLevel` property you can use to know that it needs to throw.

either way, add a little comment to make it clear this is the special handling for top level

Reply:
Oh good point! Yeah that's much more elegant.